### PR TITLE
Add exhaustive case matcher

### PR DESCRIPTION
**Pull request by peterfication**
Can be used by including `Ruby::Enum::Ecase` in an enum class. It will add a method called `ecase` that can be used to simulate a case statement that will raise an error if a case/enum value is not handled.

Fixes https://github.com/dblock/ruby-enum/issues/42

### CODE REVIEW

**Comment by peterfication** [comment]
@dblock here you go :)

File: README.md [hunk 1]
```diff
@@ -259,6 +259,42 @@ OrderState.values # ['CREATED', 'PAID']
 ShippedOrderState.values # ['CREATED', 'PAID', 'PREPARED', SHIPPED']
 ```
 
+### Exhaustive case matcher
```
Comment (dblock) [inline README.md:262]: Add to TOC.

File: README.md [hunk 2]
```diff
@@ -259,6 +259,42 @@ OrderState.values # ['CREATED', 'PAID']
...
+
+If you want to make sure that you cover all cases in a case stament, you can use the exhaustive case matcher: `Ruby::Enum::Ecase`. It will raise an error if a case/enum value is not handled or if a value is specified that's not part of the enum. This is inspired by the [Rust Pattern Syntax](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html).
```
Comment (dblock) [inline README.md:264]: add a comma, `, or if`

File: README.md [hunk 3]
```diff
@@ -259,6 +259,42 @@ OrderState.values # ['CREATED', 'PAID']
...
+  include Ruby::Enum::Ecase
+
+  define :PREPARED, 'PREPARED'
+  define :SHIPPED, 'SHIPPED'
+end
+```
+
+```ruby
+shipped_order_state = ShippedOrderState::SHIPPED
+ShippedOrderState.ecase(shipped_order_state, {
```
Comment (dblock) [inline README.md:281]: What does this return?

File: README.md [hunk 4]
```diff
@@ -259,6 +259,42 @@ OrderState.values # ['CREATED', 'PAID']
 ShippedOrderState.values # ['CREATED', 'PAID', 'PREPARED', SHIPPED']
 ```
 
+### Exhaustive case matcher
+
+If you want to make sure that you cover all cases in a case stament, you can use the exhaustive case matcher: `Ruby::Enum::Ecase`. It will raise an error if a case/enum value is not handled or if a value is specified that's not part of the enum. This is inspired by the [Rust Pattern Syntax](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html).
+
+> NOTE: This will add checks at runtime which might lead to slightly worse performance.
```
Comment (dblock) [inline README.md:266]: I am not sure this is needed. All code has performance impact. Unless you believe it's slower by a factor?

File: README.md [hunk 5]
```diff
@@ -259,6 +259,42 @@ OrderState.values # ['CREATED', 'PAID']
...
+  define :SHIPPED, 'SHIPPED'
+end
+```
+
+```ruby
+shipped_order_state = ShippedOrderState::SHIPPED
+ShippedOrderState.ecase(shipped_order_state, {
+  [ShippedOrderState::CREATED, ShippedOrderState::PAID] => -> { "order is created or paid" },
+  ShippedOrderState::PREPARED => -> { "order is prepared" },
+  ShippedOrderState::SHIPPED => -> { "order is shipped" },
```
Comment (dblock) [inline README.md:284]: Can I reference `shipped_order_state` inside the lambdas?

File: README.md [hunk 6]
```diff
@@ -259,6 +259,42 @@ OrderState.values # ['CREATED', 'PAID']
...
+### Exhaustive case matcher
+
+If you want to make sure that you cover all cases in a case stament, you can use the exhaustive case matcher: `Ruby::Enum::Ecase`. It will raise an error if a case/enum value is not handled or if a value is specified that's not part of the enum. This is inspired by the [Rust Pattern Syntax](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html).
+
+> NOTE: This will add checks at runtime which might lead to slightly worse performance.
+
+> NOTE: `:else` is a reserved keyword if you want to use `Ruby::Enum::Ecase`.
+
+```ruby
+class ShippedOrderState < OrderState
```
Comment (dblock) [inline README.md:271]: Should we use the color examples above?

File: lib/ruby-enum/enum/ecase.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    #   [Color::RED, Color::GREEN] => -> { puts "red or green" },
+    #   Color::BLUE => -> { puts "blue" },
+    # })
+    #
+    # Reserves the :else key for a default case:
+    # Color.ecase(Color::RED, {
+    #   [Color::RED, Color::GREEN] => -> { puts "red or green" },
+    #   else: -> { puts "blue" },
+    # })
+    module Ecase
```
Comment (dblock) [inline lib/ruby-enum/enum/ecase.rb:28]: Ruby doesn't have enums, so we add `Enum`, but ruby does have `case`. Does calling this `Ruby::Enum::Case`, a special kind of case, not work? Will there be ever a different `Xcase`? Finally, should it be `ECase` vs. `Ecase`?

File: lib/ruby-enum/enum/ecase.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+          results = filtered_cases.map { |_values, block| block.call }
+
+          # Return the first result if there is only one result
+          results.size == 1 ? results.first : results
+        end
+
+        private
+
+        def validate_cases(cases)
```
Comment (dblock) [inline lib/ruby-enum/enum/ecase.rb:60]: Rename to `validate_cases!` since it raises?

File: README.md (hunk 2 above)
Comment (dblock) [inline README.md:264]: Describe that all are called when multiple match.

File: spec/ruby-enum/enum/ecase_spec.rb [hunk 9]
```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+RSpec.describe Ruby::Enum::Ecase do
```
Comment (dblock) [inline spec/ruby-enum/enum/ecase_spec.rb:5]: Add tests with an empty `ecase` value, maybe an empty enum too?

**Review by dblock — CHANGES_REQUESTED** [review]
Very close! I left a bunch of comments/questions, take a look.

**Comment by peterfication** [comment]
Thanks a lot for the deep review and your comments! I appreciate it :)

File: lib/ruby-enum/enum/ecase.rb (hunk 8 above)
Comment (peterfication) [reply lib/ruby-enum/enum/ecase.rb:60]: I think the convention is that you should only provide a method with a `!` if you also provide a safe method without it, which we don't do here. Also, I think this only applies to public methods. So I would leave it as it is if that's ok with you.

File: lib/ruby-enum/enum/ecase.rb (hunk 7 above)
Comment (peterfication) [reply lib/ruby-enum/enum/ecase.rb:28]: Very good point! It's way nicer to have `Ruby::Enum::Case` and `Color.case` 👍

File: README.md (hunk 6 above)
Comment (peterfication) [reply README.md:271]: Sounds reasonable 👍

File: README.md (hunk 5 above)
Comment (peterfication) [reply README.md:284]: That's a really good idea! :)

File: README.md (hunk 4 above)
Comment (peterfication) [reply README.md:266]: The thing is that a normal case statement is pretty fast I would say. However, this enum case statement creates some lists in between and checks on them for every execution. That's why I would assume it's slower by a factor from what one would use normally. But I haven't performance tested it.

Should I still remove it? Or should I add a further explanation?

File: README.md (hunk 3 above)
Comment (peterfication) [reply README.md:281]: The return value of the block or an array of return values if multiple blocks matched. Added it to the README.

File: README.md (hunk 5 above)
Comment (peterfication) [reply README.md:285]: Implemented it.

File: README.md (hunk 4 above)
Comment (peterfication) [reply README.md:267]: I just tested it with the following code:

```ruby
  describe 'performance' do
    let(:cases) do
      {
        [test_enum::RED, test_enum::GREEN] => -> { 'red or green' },
        test_enum::BLUE => -> { 'blue' }
      }
    end

    it 'is faster than a regular case statement' do
      ruby_enum_time = Benchmark.realtime do
        1_000_000.times do
          test_enum.case(test_enum::RED, cases)
        end
      end

      case_statement_time = Benchmark.realtime do
        1_000_000.times do
          case test_enum::RED
          when test_enum::RED, test_enum::GREEN
            'red or green'
          when test_enum::BLUE
            'blue'
          end
        end
      end

      puts "ruby-enum: #{ruby_enum_time.round(4)}"
      puts "case statement: #{case_statement_time.round(4)}"

      puts "ruby-enum is #{(ruby_enum_time / case_statement_time).round(2)} times slower"
    end
  end
```

And the result is bad, so the word `slightly` should be removed or replaced with `a lot`

```ruby
ruby-enum: 2.4736
case statement: 0.0774
ruby-enum is 31.98 times slower
```

Comment (peterfication) [reply README.md:267]: I might look into ways to improve its performance ...

Comment (peterfication) [reply README.md:267]: (I used Ruby 3.1 and Ruby 3.2 is similar)

Comment (dblock) [reply README.md:267]: Nothing is free in meta programming :) Let's leave the comment! You should contribute these benchmarks as well into a new `benchmarks` subfolder with a quick how to run them README in a separate PR.

**Comment by dblock** [comment]
Make rubocop happy and we're good to go! Much thanks.

**Comment by dblock** [comment]
Oh and since it's a new feature, increment the version to 0.10.0 in this PR. I also wonder whether it's time for Ruby::Enum to graduate to 1.0. That works for me too, your call!

**Comment by peterfication** [comment]
Rubocop is failing on master. Hence, I will make the Rubocop adjustments in a separate commit.

File: README.md (hunk 4 above)
Comment (peterfication) [reply README.md]: Good idea. Done :)

**Comment by peterfication** [comment]
I had to disable and ignore some Rubocop rules. See the Rubocop related commit. I wanted to make as little changes as possible in that commit.

**Comment by peterfication** [comment]
I feel like it's a good idea to release a 1.0.0 version as it has been proven stable.

Note: the 0.9.1 version has not been tagged or released to Ruby gems yet.

**Comment by peterfication** [comment]
But maybe it makes sense to wait for https://github.com/dblock/ruby-enum/issues/41 for the 1.0.0 release.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,7 @@
+### 1.0.0
+
+* [#43](https://github.com/dblock/ruby-enum/pull/43): Add exhaustive case matcher - [@peterfication](https://github.com/peterfication).
+
 ### 0.9.1 (Next)
```
Comment (dblock) [inline CHANGELOG.md:5]: Remove/merge this with 1.0.0 above.

**Review by dblock — CHANGES_REQUESTED** [review]
For Rubocop, remove any of the manual disable and the new rules and just run `rubocop -a ; rubocop --auto-gen-config`.

**Comment by dblock** [comment]
> But maybe it makes sense to wait for #41 for the 1.0.0 release.

Are you going to take it up? :) Please!

**Comment by peterfication** [comment]
> Are you going to take it up? :) Please!

I might give it a try next week.

**Comment by dblock** [comment]
Great work @peterfication! Want to help comantain ruby-enum? Drop me your rubygems username to dblock at dblock dot org!
